tmc-langs-qmake/src/main/java/fi/helsinki/cs/tmc/langs/qmake/QmakePlugin.java

```diff
@@ -0,0 +1,309 @@
+
+    // Finds pattern 'POINT(exercise_name, 1)'
+    private static final Pattern POINT_PATTERN
+            = Pattern.compile("POINT\\(\\s*(\\w+),\\s*(\\w+)\\s*\\)\\s*;");
+    // Pattern to find comments
+    private static final Pattern COMMENT_PATTERN
+            = Pattern.compile("(^[^\"\\r\\n]*\\/\\*{1,2}.*?\\*\\/"
+                    + "|(^[^\"\\r\\n]*\\/\\/[^\\r\\n]*))", Pattern.MULTILINE | Pattern.DOTALL);
+
+    private static final Logger log = LoggerFactory.getLogger(QmakePlugin.class);
+
+    /**
+     * Creates a new QmakePlugin.
+     */
+    public QmakePlugin() {
+        super(
+                new ExerciseBuilder(),
+                new StudentFileAwareSubmissionProcessor(),
+                new StudentFileAwareZipper(),
+                new StudentFileAwareUnzipper());
+    }
+
+    @Override
+    public String getPluginName() {
+        return "qmake";
+    }
+
+    /**
+     * Resolve the exercise .pro file from exercise directory. The file should
+     * be named after the directory.
+     */
+    private Path getProFile(Path basePath) {
+        return Paths.get(basePath.toString() + "/" + basePath.getFileName() + ".pro");
+    }
+
+    @Override
+    public Optional<ExerciseDesc> scanExercise(Path path, String exerciseName) {
+        if (!isExerciseTypeCorrect(path)) {
+            log.error("Failed to scan exercise due to missing qmake project file.");
+            return Optional.absent();
+        }
+
+        Optional<ImmutableList<TestDesc>> tests = availablePoints(path);
+        if (!tests.isPresent()) {
+            log.error("Failed to scan exercise due to parsing error.");
+            return Optional.absent();
+        }
+
+        return Optional.of(new ExerciseDesc(exerciseName, ImmutableList.copyOf(tests.get())));
+    }
+
+    @Override
+    public boolean isExerciseTypeCorrect(Path path) {
+        return Files.isRegularFile(getProFile(path));
+    }
+
+    /**
+     * Gets a language specific {@link StudentFilePolicy}.
+     *
+     * <p>
+     * The project root path must be specified for the {@link StudentFilePolicy}
+     * to read any configuration files such as <tt>.tmcproject.yml</tt>.
+     * </p>
+     *
+     * @param projectPath The project's root path
+     */
+    @Override
+    protected StudentFilePolicy getStudentFilePolicy(Path projectPath) {
+        return new QmakeStudentFilePolicy(projectPath);
+    }
+
+    @Override
+    public RunResult runTests(Path path) {
+        Path shadowDir = makeShadowBuildDir(path);
+
+        try {
+            ProcessResult qmakeBuild = buildWithQmake(shadowDir);
+            if (qmakeBuild.statusCode != 0) {
+                log.error("Building project with qmake failed: {}", qmakeBuild.errorOutput);
+                return filledFailure(qmakeBuild.errorOutput);
+            }
+        } catch (Exception e) {
+            log.error("Building project with qmake failed", e);
+            throw new RuntimeException(e);
```

Comment:
What do you mean by custom (checked) exception? Like `QmakeBuildException` and `MakeBuildException` etc? Would it be helpful to include stdout/stderr output to exceptions from ProcessResult?

Reply:
Yeah, something like that. For example the Java plugins throw a [`fi.helsinki.cs.tmc.langs.java.exception.TestRunnerException `](https://github.com/testmycode/tmc-langs/blob/master/tmc-langs-java/src/main/java/fi/helsinki/cs/tmc/langs/java/exception/TestRunnerException.java) in a somewhat analogous situation.

I don't know how familiar you are with checked vs unchecked exceptions, but here's a short primer that also represents a pretty sensible stance on when to use unchecked exceptions: https://docs.oracle.com/javase/tutorial/essential/exceptions/runtime.html.  

Unchecked exceptions (`RuntimeException`, `Error` and their subclasses) are a bit controversial and Google will find a plethora of extreme opinions. The standard stance seems to be that checked exceptions are used when the exception is such that the code calling your code can be expected to recover and unchecked exceptions are used for behavior that is clearly a bug in the code calling your code.

As a concrete example, if you call a method that excepts an instance of `Thing` as it's parameter with `null` argument, it's OK for that method to throw a `NullPointException`, which is unchecked and not part of the method signature, because *your code* is using the method wrong. In other words, the bug is not in the method but the code calling it. Similarly it's fine for `divide(0, 0)` to throw an unchecked `ArithmeticException` since clearly the code calling the function is doing it wrong.

At the same time, we want to throw a checked exception like `IOException` when a disk read fails, as that's something that the calling code can be expected to -- or atleast attempt to -- recover from.

If that made no sense, hit me up on Slack :)

---

As for including the stdout/stderr, it's probably unnecessary to include them in the exception itself, but logging them is probably a sensible thing to do. At least log the stderr when things go wrong. Essentially consider this: what information would *you* want to have when you debug a problem happening on someone else's computer and you only have the log files? :)

Reply:
Thank you for the lengthy explanation, tried searching for the concept and found the controversy (that was really the problem here). Will change unchecked exceptions to checked ones wherever they seem to be desired.

Should be logging stderr currently at every point when something goes wrong.